mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java

```diff
@@ -0,0 +1,571 @@
+    }
+
+    /**
+     * Compute platforms pass down filters to data sources.
+     * If the data source cannot apply some filters, or only
+     * partially applies the filter, it will return the
+     * residual filter back. If the platform
+     * can correctly apply the residual filters, then it
+     * should call this api. Otherwise the current
+     * api will throw an exception if the passed in
+     * filter is not completely satisfied. Note: This
+     * does not apply to standalone MR application
+     */
+    public ConfigBuilder platformAppliesFilterResiduals() {
+      conf.setBoolean(PLATFORM_APPLIES_FILTER_RESIDUALS, true);
+      return this;
+    }
+  }
+
+  @Override
+  public List<InputSplit> getSplits(JobContext context) {
+    if (splits != null) {
+      LOG.info("Returning cached splits: {}", splits.size());
+      return splits;
+    }
+
+    Configuration conf = context.getConfiguration();
+    Table table = findTable(conf);
+    TableScan scan = table.newScan()
+                          .caseSensitive(conf.getBoolean(CASE_SENSITIVE, true));
+    long snapshotId = conf.getLong(SNAPSHOT_ID, -1);
+    if (snapshotId != -1) {
+      scan = scan.useSnapshot(snapshotId);
+    }
+    long asOfTime = conf.getLong(AS_OF_TIMESTAMP, -1);
+    if (asOfTime != -1) {
+      scan = scan.asOfTime(asOfTime);
+    }
+    long splitSize = conf.getLong(SPLIT_SIZE, 0);
+    if (splitSize > 0) {
+      scan = scan.option(TableProperties.SPLIT_SIZE, String.valueOf(splitSize));
+    }
+    String schemaStr = conf.get(READ_SCHEMA);
+    if (schemaStr != null) {
+      scan.project(SchemaParser.fromJson(schemaStr));
+    }
+
+    // TODO add a filter parser to get rid of Serialization
+    Expression filter = SerializationUtil.deserializeFromBase64(conf.get(FILTER_EXPRESSION));
+    if (filter != null) {
+      scan = scan.filter(filter);
+    }
+
+    splits = Lists.newArrayList();
+    try (CloseableIterable<CombinedScanTask> tasksIterable = scan.planTasks()) {
+      tasksIterable.forEach(task -> {
+        checkResiduals(conf, task);
+        splits.add(new IcebergSplit(conf, task));
+      });
+    } catch (IOException e) {
+      throw new RuntimeIOException(e, "Failed to close table scan: %s", scan);
+    }
+
+    return splits;
+  }
+
+  private static void checkResiduals(Configuration conf, CombinedScanTask task) {
+    boolean platformAppliesFilter = conf.getBoolean(PLATFORM_APPLIES_FILTER_RESIDUALS, false);
+    //TODO remove the check on dataModel once we start supporting
+    // residual evaluation for Iceberg Generics in InputFormat
+    InMemoryDataModel dataModel = conf.getEnum(IN_MEMORY_DATA_MODEL, InMemoryDataModel.DEFAULT);
+    if (dataModel == InMemoryDataModel.DEFAULT || !platformAppliesFilter) {
+      task.files().forEach(fileScanTask -> {
+        Expression residual = fileScanTask.residual();
+        if (residual != null && !residual.equals(Expressions.alwaysTrue())) {
+          throw new RuntimeException(
```

Comment:
How about `UnsupportedOperationException` since this will eventually be implemented?

Reply:
Well, this would only be implemented for GENERIC data model, not for Pig and Hive. Thoughts on keeping the same exception or do u think we should implement residual eval for Pig and Hive data models?

Reply:
My rationale was that we would accept a patch to add residual filtering for Hive and Pig, so why not consider it a future feature? Should someone choose to contribute it I think it would be valuable.

Reply:
OK that make sense

Reply:
Kept as `UnsupportedException`